proto-lens/src/Data/ProtoLens/Encoding.hs

```diff
@@ -137,42 +157,51 @@ buildMessage msg = foldMap putTaggedValue (messageToTaggedValues msg)
 -- | Encode a message as a sequence of key-value pairs.
 messageToTaggedValues :: Message msg => msg -> [TaggedValue]
 messageToTaggedValues msg = mconcat
-    [ map (TaggedValue t) (messageFieldToVals fieldDescr msg)
+    [ messageFieldToVals t fieldDescr msg
     | (Tag t, fieldDescr) <- Map.toList (fieldsByTag descriptor)
     ]
 
-messageFieldToVals :: FieldDescriptor msg -> msg -> [WireValue]
-messageFieldToVals (FieldDescriptor _ typeDescriptor accessor) msg =
-    case fieldWireType typeDescriptor of
-        FieldWireType wt convert _ -> case accessor of
+messageFieldToVals :: Int -> FieldDescriptor a -> a -> [TaggedValue]
+messageFieldToVals tag (FieldDescriptor _ typeDescriptor accessor) msg =
+    let
+        embed src
+            = case fieldWireType typeDescriptor of
+                FieldWireType wt convert _
+                    -> [TaggedValue tag $ WireValue wt
+                                            (convert src)]
+                GroupFieldType
+                    -> TaggedValue tag (WireValue StartGroup ())
+                            : messageToTaggedValues src
+                                ++ [TaggedValue tag $ WireValue EndGroup ()]
+        embedPacked src
+            = case fieldWireType typeDescriptor of
+                GroupFieldType -> fail "packged GroupFieldType"
+                FieldWireType wt convert _ -> let
+                    v = L.toStrict $ toLazyByteString
+                        $ mconcat [putWireValue wt (convert x) | x <- src]
+                    in [TaggedValue tag $ WireValue Lengthy v]
+    in case accessor of
             PlainField d f
                 | Optional <- d, src == fieldDefault -> []
-                | otherwise -> [WireValue wt (convert src)]
+                | otherwise -> embed src
               where src = msg ^. f
             OptionalField f -> case msg ^. f of
-                Just src -> [WireValue wt (convert src)]
-                _ -> mempty
+                                Just src -> embed src
+                                _ -> mempty
             RepeatedField Unpacked f
-                -> [ WireValue wt (convert src)
-                   | src <- toList (msg ^. f)
-                   ]
+                -> mconcat [embed src | src <- toList (msg ^. f)]
```

Comment:
It turns out foldMap doesn't fuse the intermediate list like the previous code did.  Switched to concatMap which produces similar code to before.